File: SBiswas/Problems/Milestone1/Problem17.cpp
/* Mode of BST tree*/

#include <iostream>
#include <vector>
#include <unordered_map>


using namespace std;

class TreeNode {
    
    public:
    int data;
    TreeNode* left;
    TreeNode* right;
    
    TreeNode(int val)
    {
        data = val;
        left = nullptr;
        right = nullptr;
    }
    
};

class Solution {
public:
    void inorder(TreeNode* root, unordered_map<int,int>& node_count)
    {
        if(!root)
        {
            return;
        }

        inorder(root->left, node_count);

        if(node_count.find(root->data) == node_count.end())
        {
            node_count[root->data] = 0;
        }
        node_count[root->data] += 1;

        inorder(root->right, node_count);

    }
    vector<int> findMode(TreeNode* root) {
        
        unordered_map<int,int> node_count;
        inorder(root, node_count);

        vector<int> ans;
        int max_freq=0;
        for(auto i : node_count)
        {
            max_freq = max(max_freq, i.second);
        }

        for(auto i : node_count)
        {
            if(i.second == max_freq)
            {
                ans.push_back(i.first);
            }
        }

        return ans;
    }
};

File: Hritwik/Problems/Milestone1/Problem17.py
# Definition for a binary tree node.
from collections import deque
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:   

    def findMode(self, root): 
        if not root:
            return []
        hash_map = {}
        queue = deque([root])
        while queue:
            level = []
            len_of_queue = len(queue)
            for _ in range(len_of_queue):
                node = queue.popleft()
                if node.val in hash_map:
                    hash_map[node.val]+=1
                else:
                    hash_map[node.val]=1
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
        max_value = max(hash_map.values())
        mode_keys = [key for key, value in hash_map.items() if value == max_value]
               
        return mode_keys

solution = Solution()

# Create a binary tree to traverse through it in an inorder traversal manner
#
#         Root   2
#              /    \
#             2      5  
#                   /  \
#                  5    7
#
root = TreeNode(2)
root.left = TreeNode(2)
root.right = TreeNode(5)
root.right.left = TreeNode(5)
root.right.right = TreeNode(7)


response = solution.findMode(root)
print(f'The mode(s) of the BST with the given root is/are. -> {response}')

